### PR TITLE
suit/examples: add gpio callback trigger

### PR DESCRIPTION
**Pull request by fjmolinas**
### Contribution description

This PR adds a cb that triggers an update when a button is executed. This resembles a little more how a real deployment would work, devices would not be waiting for notify messages but periodically check the fw update server. Going from a `gpio` callback to an `xtimer_set()` based callback would be direct. 

### Testing procedure

On a board supporting `periph_gpio_irq `

- start `aiocoap-fileserver`, if README is followed you should be able to do:

```
exec aiocoap-fileserver coaproot/
```

- setup network

```
sudo dist/tools/ethos/setup_network.sh riot0 ${LOCAL_IP6_PREFIX}::/64
```

- bootstrap device:
```
SUIT_COAP_SERVER=[fd00:dead:beef::1] make -C examples/suit_update/ BOARD=samr21-xpro flash -j term
```

- Publish new firmware
```
SUIT_COAP_SERVER=[fd00:dead:beef::1] make -C examples/suit_update suit/publish -j4
```

- Press button, it will start updating:

```
main(): This is RIOT! (Version: 2020.01-devel-134-g0bcb6-pr_suit_update_trigger)
RIOT SUIT update example application
running from slot 0
slot start addr = 0x1000
Image magic_number: 0x544f4952
Image Version: 0x5da9714e
Image start address: 0x00001100
Header chksum: 0x5db07d99

suit_coap: started.
Starting the shell
> Button preseed! Triggering suit Update! 
suit_coap: trigger received
suit_coap: downloading "coap://[fd00:dead:beef::1]/fw/samr21-xpro/suit_update-riot.suitv4_signed.latest.bin"
suit_coap: got manifest with size 545
suit: verifying manifest signature...
```
### Issues/PRs references

Depends on #12471

### CODE REVIEW

File: examples/suit_update/main.c [hunk 1]
```diff
@@ -19,20 +19,50 @@
...
+}
 
-/* import "ifconfig" shell command, used for printing addresses */
-extern int _gnrc_netif_config(int argc, char **argv);
+/* assuming that first button is always BTN0 */
+#if defined(MODULE_PERIPH_GPIO_IRQ) && defined(BTN0_PIN)
+static void cb(void *arg)
+{
+    (void) arg;
+    printf("Button presed! Triggering suit Update! \n");
```
Comment (aabadie) [inline examples/suit_update/main.c]:
```suggestion
    printf("Button pressed! Triggering suit Update! \n");
```

**Comment by fjmolinas** [comment]
@aabadie no longer waiting for another PR, I updated the test procedure.

File: examples/suit_update/Makefile [hunk 2]
```diff
@@ -31,6 +31,9 @@ USEMODULE += gnrc_icmpv6_echo
 USEMODULE += shell
 USEMODULE += shell_commands
 
+# Optional feature to trigger update though gpio
```
Comment (aabadie) [inline examples/suit_update/Makefile]:
```suggestion
# Optional feature to trigger update through gpio
```

File: examples/suit_update/Makefile [hunk 3]
```diff
@@ -82,6 +85,9 @@ TEST_EXTRA_FILES += $(SLOT_RIOT_ELFS) $(SUIT_SEC) $(SUIT_PUB)
 
 include $(RIOTBASE)/Makefile.include
 
+# Export default path to fetch an update
+CFLAGS += -DSUIT_MANIFEST_RESOURCE=\"$(SUIT_COAP_ROOT)/$(SUIT_NOTIFY_MANIFEST)\"
```
Comment (aabadie) [inline examples/suit_update/Makefile]: This change is quite interesting. I'm thinking that an improvement of the new `suit` shell command could also be based on that: calling the `suit` without parameter would trigger the update based on the pre-provisionned manifest resource. What do you think ?
Not needed for this PR though.

File: examples/suit_update/main.c [hunk 4]
```diff
@@ -54,6 +58,16 @@ static void *_nanocoap_server_thread(void *arg)
     return NULL;
 }
 
+/* assuming that first button is always BTN0 */
+#if defined(MODULE_PERIPH_GPIO_IRQ) && defined(BTN0_PIN)
+static void cb(void *arg)
+{
+    (void) arg;
+    printf("Button pressed! Triggering suit Update! \n");
```
Comment (aabadie) [inline examples/suit_update/main.c]:
```suggestion
    printf("Button pressed! Triggering suit update!\n");
```

File: examples/suit_update/main.c [hunk 5]
```diff
@@ -54,6 +58,16 @@ static void *_nanocoap_server_thread(void *arg)
...
+    suit_coap_trigger((uint8_t*) SUIT_MANIFEST_RESOURCE, sizeof(SUIT_MANIFEST_RESOURCE));
```
Comment (aabadie) [inline examples/suit_update/main.c]:
```suggestion
    suit_coap_trigger((uint8_t *)SUIT_MANIFEST_RESOURCE, sizeof(SUIT_MANIFEST_RESOURCE));
```

File: examples/suit_update/main.c [hunk 6]
```diff
@@ -74,6 +88,12 @@ int main(void)
         printf("[FAILED] You're not running riotboot\n");
     }
 
+#if defined(MODULE_PERIPH_GPIO_IRQ) && defined(BTN0_PIN)
+    /* initiates a button to manually trigger an update */
```
Comment (aabadie) [inline examples/suit_update/main.c]:
```suggestion
    /* initialize a button to manually trigger an update */
```

File: makefiles/suit.v4.inc.mk [hunk 7]
```diff
@@ -10,8 +10,9 @@ SUIT_MANIFEST_LATEST ?= $(BINDIR_APP)-riot.suitv4.latest.bin
 SUIT_MANIFEST_SIGNED ?= $(BINDIR_APP)-riot.suitv4_signed.$(APP_VER).bin
 SUIT_MANIFEST_SIGNED_LATEST ?= $(BINDIR_APP)-riot.suitv4_signed.latest.bin
 
+
```
Comment (aabadie) [inline makefiles/suit.v4.inc.mk]: unrelated ?

File: examples/suit_update/Makefile (hunk 3 above)
Comment (aabadie) [reply examples/suit_update/Makefile]: Another minor code comment: this line could be moved before include `$(RIOTBASE)/Makefile.include`

**Comment by aabadie** [comment]
Testing this PR now and it works like a charm, awesome!

**Comment by aabadie** [comment]
The default automatic test is also working

File: examples/suit_update/Makefile (hunk 3 above)
Comment (fjmolinas) [reply examples/suit_update/Makefile]: How about I move it next to `peripg_gpio_irq`:
```
# Optional feature to trigger suit update through gpio callback
FEATURES_OPTIONAL += periph_gpio_irq
# Default COAP manifest resource location when fetched though gpio callback
CFLAGS += -DSUIT_MANIFEST_RESOURCE=\"$(SUIT_COAP_ROOT)/$(SUIT_NOTIFY_MANIFEST)\"
```

Comment (aabadie) [reply examples/suit_update/Makefile]: With a blank line between the 2

Comment (fjmolinas) [reply examples/suit_update/Makefile]: Ups. No it actually can't be moved higher since it depends on the value of `SUIT_COAP_ROOT` and `SUIT_NOTIFY_MANIFEST`.
